packages/mobx/src/core/reaction.ts

```diff
@@ -90,6 +94,16 @@ export class Reaction implements IDerivation, IReactionPublic {
      * internal, use schedule() if you intend to kick off a reaction
      */
     runReaction_() {
+        if (!this.scheduledRunReaction_) {
+            this.scheduledRunReaction_ = true
+            this.scheduler(() => {
+                this.scheduledRunReaction_ = false
+                this.runReactionImpl()
+            })
+        }
+    }
+
+    private runReactionImpl() {
```

Comment:
These are the core changes - `Reaction` dictates when it should `runReactionImpl` based on the passed in scheduler.